### PR TITLE
haskellPackages.mkDerivation: New `intermediates` output

### PR DESCRIPTION
**Pull request by Gabriella439**
This adds a new `intermediates` output that can be used to accelerate a Haskell build by importing build products exported from a prior similar build.

The motivation for this is explained by the following post:

https://www.haskellforall.com/2022/12/nixpkgs-support-for-incremental-haskell.html

This is a subset of the changes from #204020 that I believe are safe to merge while waiting for https://github.com/NixOS/nix/pull/7362 to be reviewed.  I've incorporated outstanding feedback from the last PR into this one.

I structured this to be a hash-preserving changes for Haskell packages that don't use the feature, so this won't trigger unnecessary rebuilds or breakage.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sternenseemann** [comment]
> This is a subset of the changes from #204020 that I believe are safe to merge while waiting for [NixOS/nix#7362](https://github.com/NixOS/nix/pull/7362) to be reviewed. I've incorporated outstanding feedback from the last PR into this one.

Without that Nix change, this is hardly useful, though?

**Comment by Gabriella439** [comment]
Not necessarily.  There are other ways to thread the full build to the incremental build, albeit with a worse UX.  See, for example: https://felixspringer.xyz/homepage/blog/incrementalHaskellBuildsWithNix

**Comment by cdepillabout** [comment]
@Gabriella439 I think the full functionality in https://github.com/NixOS/nixpkgs/pull/212107 could be pretty neat, but I'm not sure how obvious it would be to use.

Would you be willing to add a test for this functionality to `pkgs/test/haskell/` (if that makes sense)?  That might also work as an example of how the new `intermediates` output could be used.

**Comment by 9999years** [comment]
@cdepillabout I've added a test to show how to use the API. For now, how to manage the `previousIntermediates` is left as an exercise to the reader. But once https://github.com/NixOS/nix/pull/7362 is merged (we're working on getting a review from the Nix team) we'll be able to merge in the rest of https://github.com/NixOS/nixpkgs/pull/204020 (with a smaller diff).

File: pkgs/test/haskell/incremental/default.nix
Comment (cdepillabout) [inline pkgs/test/haskell/incremental/default.nix:31]: Would you consider adding you guys as maintainers to this test?

Or potentially adding Mercury to https://github.com/NixOS/nixpkgs/blob/b03ac42b0734da3e7be9bf8d94433a5195734b19/maintainers/team-list.nix, and then making mercury the maintainer?

**Review by cdepillabout — APPROVED** [review]
LGTM.  Nice test!

cc @sternenseemann about this as well, since I think he had a few concerns previously

File: pkgs/test/haskell/incremental/default.nix
Comment (9999years) [reply pkgs/test/haskell/incremental/default.nix:31]: Hm, would that integrate correctly with GitHub? I'm not sure GitHub would be able to correctly `@` us as reviewers.

Comment (9999years) [reply pkgs/test/haskell/incremental/default.nix]: Done. Can you check that I added the maintainers correctly?

**Comment by mweinelt** [comment]
I'm locking and closing this pull request, because through the rebase you inadvertently requested many maintainers for review, which subscribed them to update notifications, resulting in unnecessary spam inside everyone's inbox. While we can remove their review requests, we sadly cannot unsubscribe anyone.

Please create a new pull request and for the next time remember to set your PR to draft status before rebasing. In draft status, you can preview the list of maintainers that are about to be requested for review, which allows you to sidestep this issue.

**Pull request closed without merging**
